### PR TITLE
CRUD of Threads

### PR DESCRIPTION
**Pull request by reza-newgen**

### CODE REVIEW

File: src/main/java/edu/miu/cs545/project/controller/ThreadPostController.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Repository;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+@RestController
+@RequestMapping("/api/v1/threadposts")
+public class ThreadPostController {
+
```
Comment (darkhanzhursin) [inline src/main/java/edu/miu/cs545/project/controller/ThreadPostController.java]: use abstract CrudController class

File: src/main/java/edu/miu/cs545/project/dto/ThreadPostDto.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+import java.time.LocalDateTime;
+
+
+@Setter
+@Getter
+@AllArgsConstructor
+public class ThreadPostDto {
+
+    private Long id;
```
Comment (darkhanzhursin) [inline src/main/java/edu/miu/cs545/project/dto/ThreadPostDto.java:10]: use Id annotation and GeneratedValue

File: src/main/java/edu/miu/cs545/project/service/impl/ThreadPostServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+            }
+
+            threadPostRepository.deleteById(id);
+        }catch (Exception e){
+        throw new RuntimeException(HttpStatus.EXPECTATION_FAILED.toString());
+    }
+    }
+
+
+    private ThreadPostDto convertToDto(ThreadPost threadPost) {
```
Comment (darkhanzhursin) [inline src/main/java/edu/miu/cs545/project/service/impl/ThreadPostServiceImpl.java]: use MapStruct library, exclude this method to other class

File: src/main/java/edu/miu/cs545/project/service/impl/ThreadPostServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+        return new ThreadPostDto(
+                threadPost.getId(),
+                threadPost.getTitle(),
+                threadPost.getCreatedAt(),
+                threadPost.getCategory() != null ? threadPost.getCategory().getId() : null,
+                threadPost.getUser() != null ? threadPost.getUser().getId() : null
+        );
+    }
+
+    private ThreadPost convertToEntity(ThreadPostDto threadPostDto) {
```
Comment (darkhanzhursin) [inline src/main/java/edu/miu/cs545/project/service/impl/ThreadPostServiceImpl.java]: same

File: src/main/java/edu/miu/cs545/project/service/impl/ThreadPostServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+import org.springframework.transaction.annotation.Transactional;
+
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+@Service
+public class ThreadPostServiceImpl implements ThreadPostService {
```
Comment (darkhanzhursin) [inline src/main/java/edu/miu/cs545/project/service/impl/ThreadPostServiceImpl.java]: use CrudServiceImpl abstract class for boilerplates

File: src/main/java/edu/miu/cs545/project/service/ThreadPostService.java [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+import java.util.List;
+import java.util.Optional;
+
+public interface ThreadPostService {
+    ResponseEntity<?> save(ThreadPostDto threadPostDto);
+    ThreadPostDto  getThreadPostById(Long id);
+    List<ThreadPostDto> getAllThreadPost();
+    ThreadPostDto  updateThreadPost(ThreadPostDto threadPostDto);
+    void deleteThreadPost(Long id);
```
Comment (darkhanzhursin) [inline src/main/java/edu/miu/cs545/project/service/ThreadPostService.java]: use CrudService

**Review by darkhanzhursin — CHANGES_REQUESTED** [review]
please, see other PRs

**Review by darkhanzhursin — APPROVED** [review]

File: src/main/java/edu/miu/cs545/project/controller/ThreadPostController.java [hunk 7]
```diff
@@ -0,0 +1,20 @@
+package edu.miu.cs545.project.controller;
+
+import edu.miu.cs545.project.model.entity.ThreadPost;
+import edu.miu.cs545.project.service.impl.ThreadPostServiceImpl;
+
+import io.swagger.v3.oas.annotations.tags.Tag;
+import org.springframework.web.bind.annotation.*;
+
+
```
Comment (nabil-miu) [inline src/main/java/edu/miu/cs545/project/controller/ThreadPostController.java]: could you remove blank lines

File: src/main/java/edu/miu/cs545/project/repository/ThreadPostRepository.java [hunk 8]
```diff
@@ -0,0 +1,11 @@
+package edu.miu.cs545.project.repository;
+
+import edu.miu.cs545.project.model.entity.ThreadPost;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+import java.util.Optional;
+
+@Repository
+public interface ThreadPostRepository extends JpaRepository<ThreadPost,Long>{
+  Optional<ThreadPost> findByTitle(String threadTitle);
```
Comment (nabil-miu) [inline src/main/java/edu/miu/cs545/project/repository/ThreadPostRepository.java]: findByTitleLike

File: src/main/java/edu/miu/cs545/project/model/entity/Category.java [hunk 9]
```diff
@@ -5,6 +5,8 @@
 import lombok.Data;
 import lombok.Setter;
 
+import java.util.Set;
```
Comment (nabil-miu) [inline src/main/java/edu/miu/cs545/project/model/entity/Category.java]: remove

File: src/main/java/edu/miu/cs545/project/controller/ThreadPostController.java [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+import edu.miu.cs545.project.model.entity.ThreadPost;
+import edu.miu.cs545.project.service.impl.ThreadPostServiceImpl;
+import io.swagger.v3.oas.annotations.tags.Tag;
+import org.springframework.web.bind.annotation.*;
+
+@RestController
+@RequestMapping("/api/v1/thread-posts")
+@Tag(name = "ThreadPost", description = "Thread Post API")
+public class ThreadPostController extends CrudController<ThreadPost,Long>{
+    public ThreadPostController(ThreadPostServiceImpl threadPostService) {
```
Comment (nabil-miu) [inline src/main/java/edu/miu/cs545/project/controller/ThreadPostController.java]: ThreadPostService

**Review by nabil-miu — APPROVED** [review]

**Pull request merged**